src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+    (is-eq char 0x62) ;; b
+    (is-eq char 0x63) ;; c
+    (is-eq char 0x64) ;; d
+    (is-eq char 0x65) ;; e
+    (is-eq char 0x66) ;; f
+    (is-eq char 0x67) ;; g
+    (is-eq char 0x68) ;; h
+    (is-eq char 0x69) ;; i
+    (is-eq char 0x6a) ;; j
+    (is-eq char 0x6b) ;; k
+    (is-eq char 0x6c) ;; l
+    (is-eq char 0x6d) ;; m
+    (is-eq char 0x6e) ;; n
+    (is-eq char 0x6f) ;; o
+    (is-eq char 0x70) ;; p
+    (is-eq char 0x71) ;; q
+    (is-eq char 0x72) ;; r
+    (is-eq char 0x73) ;; s
+    (is-eq char 0x74) ;; t
+    (is-eq char 0x75) ;; u
+    (is-eq char 0x76) ;; v
+    (is-eq char 0x77) ;; w
+    (is-eq char 0x78) ;; x
+    (is-eq char 0x79) ;; y
+    (is-eq char 0x7a))) ;; z
+
+(define-private (is-vowel (char (buff 1)))
+  (or 
+    (is-eq char 0x61) ;; a
+    (is-eq char 0x65) ;; e
+    (is-eq char 0x69) ;; i
+    (is-eq char 0x6f) ;; o
+    (is-eq char 0x75) ;; u
+    (is-eq char 0x79))) ;; y
+
+(define-private (is-special-char (char (buff 1)))
+  (or 
+    (is-eq char 0x2d) ;; -
+    (is-eq char 0x5f))) ;; _
+
+(define-private (is-char-valid (char (buff 1)))
+  (or 
+    (is-lowercase-alpha char)
+    (is-digit char)
+    (is-special-char char)))
+
+(define-private (is-nonalpha (char (buff 1)))
+  (or 
+    (is-digit char)
+    (is-special-char char)))
+
+(define-private (has-vowels-chars (name (buff 16)))
+  (> (len (filter is-vowel name)) u0))
+
+(define-private (has-nonalpha-chars (name (buff 16)))
+  (> (len (filter is-nonalpha name)) u0))
+
+(define-private (has-invalid-chars (name (buff 19)))
+  (< (len (filter is-char-valid name)) (len name)))
+
+(define-private (compute-name-price (name (buff 16))
+                                    (price-function (tuple (buckets (list 16 uint)) (base uint) (coeff uint) (nonalpha-discount uint) (no-vowel-discount uint))))
+  (let (
+    (exponent (get-exp-at-index (get buckets price-function) (min u15 (- (len name) u1))))
+    (no-vowel-discount (if (not (has-vowels-chars name)) (get no-vowel-discount price-function) u1))
+    (nonalpha-discount (if (has-nonalpha-chars name) (get nonalpha-discount price-function) u1)))
+    (*
+      (/
+        (*
+          (get coeff price-function)
+          (pow (get base price-function) exponent))
+        (max nonalpha-discount no-vowel-discount))
+      u10))) ;; 10 = name_cost (100) * "old_price_multiplier" (0.1) - todo(ludo): sort this out.
+
+(define-private (is-name-lease-expired? (namespace (buff 19)) (name (buff 16)))
+  (let (
+    (namespace-props (unwrap! 
+      (map-get? namespaces ((namespace namespace))) 
+      (err ERR_NAMESPACE_NOT_FOUND)))
+    (name-props (unwrap! 
+      (map-get? name-properties ((namespace namespace) (name name))) 
+      (err ERR_NAME_NOT_FOUND))))
+    (if (and (is-none (get registered-at name-props)) (is-some (get imported-at name-props)))
+      (ok false) ;; The name was imported and not launched - not subject to expiration, however should they expire if the namespace expire? - if so, todo(ludo)
```

Comment:
If a namespace was not launched, then all imported names expire and can be re-registered if the namespace is re-created.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/a16f8c8d20520a56095f9133093df8ed97ef81f4